src/bindings/python/flux/job/journal.py

```diff
@@ -0,0 +1,259 @@
+    or the registered callback will return None.
+
+    When the consumer is first started, historical data (events in the
+    past) will be sent from the job manager unless *full* is set to False.
+    These events are stored until all historical events are processed,
+    then are time ordered before returning them via :func:`poll` or to
+    the callback registered by :func:`set_callback`.
+
+    To avoid processing previously seen events with a new instance of this
+    class, the timestamp of the newest processed event can be passed via the
+    *since* parameter. Timestamps should be unique so :func:`poll` or the
+    callback will start with the newest event after the *since* timestamp.
+
+    When *full* is True, the job manager sends a sentinel event in
+    the journal stream to demarcate the transition between history
+    and current events. If *include_sentinel* is set to True,
+    then an empty event will be returned by :func:`poll` or the
+    registered callback to represent the sentinel. An empty event can
+    be compared to :data:`flux.job.journal.SENTINEL_EVENT` or by using the
+    :func:`JournalEvent.is_empty` method. The default behavior is to
+    suppress the sentinel.
+
+    """
+
+    def __init__(self, flux_handle, full=True, since=0.0, include_sentinel=False):
+        self.handle = flux_handle
+        self.backlog = deque()
+        self.full = full
+        self.since = since
+        self.rpc = None
+        self.cb = None
+        self.cb_args = []
+        self.cb_kwargs = {}
+        self.processing_inactive = self.full
+        self.include_sentinel = include_sentinel
+
+    def __sort_backlog(self):
+        self.processing_inactive = False
+        self.backlog = deque(sorted(self.backlog, key=lambda x: x.timestamp))
+        if self.include_sentinel:
+            self.backlog.append(SENTINEL_EVENT)
+
+    def __enqueue_response(self, resp):
+        if resp is None:
+            # End of data, enqueue None:
+            self.backlog.append(None)
+            return
+
+        jobid = resp["id"]
+        jobspec = resp.get("jobspec")
+        R = resp.get("R")
+        for entry in resp["events"]:
+            event = JournalEvent(jobid, entry)
+            if event.timestamp > self.since:
+                if event.name == "submit":
+                    event.jobspec = jobspec or None
+                elif event.name == "alloc":
+                    event.R = R or None
+                self.backlog.append(event)
+
+    def __next_event(self):
+        return self.backlog.popleft()
+
+    def start(self):
+        """Start the stream of events by sending a request to the job manager
+
+        This function sends the job-manager.events-journal RPC to the
+        job manager. It must be called to start the stream of events.
+        """
+        self.rpc = self.handle.rpc(
+            "job-manager.events-journal", {"full": self.full}, 0, FLUX_RPC_STREAMING
```

Comment:
Oh, one thing I just thought of -- take it or leave it: the journal RPC gives you the ability to subscribe to [only certain events](https://github.com/flux-framework/flux-core/blob/master/src/modules/job-manager/journal.c#L16), such as start and finish, maybe we want to expose that option to `JournalConsumer` as well?

I think most services are going to be interested in the full event log, though.

Reply:
Ah, ok, for another time :)